aws/lambda/log-classifier/src/engine.rs

```diff
@@ -44,3 +46,312 @@ pub fn evaluate_ruleset(ruleset: &RuleSet, log: &Log) -> Option<Match> {
         .flat_map(|rule| evaluate_rule(rule, log))
         .max_by(|a, b| a.rule.priority.cmp(&b.rule.priority))
 }
+
+/// Evaluates a ruleset against all lines of a log, returning matching line numbers.
+///
+/// # Arguments
+///
+/// * `ruleset` - A reference to the RuleSet to evaluate.
+/// * `log` - A reference to the Log to evaluate against.
+///
+/// # Returns
+///
+/// A vector of line numbers that match any rule in the ruleset.
+/// Note: Each rule in ruleset can only match 1 line which is the last line that matches the rule.
+pub fn evaluate_ruleset_all_lines(ruleset: &RuleSet, log: &Log) -> Vec<usize> {
+    let matches: Vec<Match> = ruleset
+        .rules
+        .par_iter()
+        .flat_map(|rule| evaluate_rule(rule, log))
+        .collect();
+    // combine captures into a single vector
+    let mut lines = Vec::new();
+    for m in matches {
+        lines.push(m.line_number);
+    }
+    lines
+}
+
+/// Merges line numbers into chunks based on specified size constraints.
+///
+/// # Arguments
+///
+/// * `log` - A reference to the Log structure containing the lines.
+/// * `line_numbers` - A vector of line numbers to be chunked.
+/// * `min_chunk_size` - The minimum size of each chunk.
+/// * `max_chunk_size` - The maximum size of each chunk.
+///
+/// # Returns
+///
+/// A vector of tuples, where each tuple contains the start and end line numbers of a chunk.
+///
+/// # Panics
+///
+/// This function will panic if:
+/// * `min_chunk_size` is greater than or equal to `max_chunk_size`.
+/// * Any line number is less than 1 or greater than the number of lines in the log.
+fn get_line_number_chunks(
+    log: &Log,
+    line_numbers: Vec<usize>,
+    min_chunk_size: usize,
+    max_chunk_size: usize,
+) -> Vec<(usize, usize)> {
+    // throw user error if min_chunk_size is greater than max_chunk_size
+    if min_chunk_size >= max_chunk_size {
```

Comment:
Should we really panic on `min_chunk_size == max_chunk_size` ?

Reply:
In theory no, but I'm not sure why you'd want to support that functionality from a user perspective, and I was running into some problems when testing it. I think it's fine without it.

Reply:
>  I was running into some problems when testing it.

I think this indicates a bug in the code. The correct code should handle this corner case naturally.

Reply:
I think this is logically no longer a problem due to the new api